include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -38,12 +38,14 @@ namespace impl
     template <typename T>
     uint64_t promoteToUint64(T val)
     {
-        using AsFloat = unsigned_integer_of_size<sizeof(T)>;
+        using AsFloat = typename float_of_size<sizeof(T)>::type;
         uint64_t asUint = ieee754::impl::castToUintType(val);
 
-        const uint64_t sign = (uint64_t(ieee754::getSignMask<T>()) | asUint) << (sizeof(uint64_t) - sizeof(T) - 2);
-        const uint64_t exp = (uint64_t(ieee754::getExponentMask<T>()) | asUint) << (ieee754::getMantissaBitCnt<uint64_t>() - ieee754::getMantissaBitCnt<T>() + ieee754::getExponentBitCnt<T>());
-        const uint64_t mantissa = (uint64_t(ieee754::getMantissaMask<T>()) | asUint) << (ieee754::getMantissaBitCnt<uint64_t>() - ieee754::getMantissaBitCnt<T>());
+        const uint64_t sign = (uint64_t(ieee754::getSignMask<AsFloat>()) & asUint) << (sizeof(float64_t) - sizeof(T));
+        const int64_t newExponent = ieee754::extractExponent(val) + ieee754::getExponentBias<float64_t>();
+
+        const uint64_t exp = (uint64_t(ieee754::extractExponent(val)) + ieee754::getExponentBias<float64_t>()) << (ieee754::getMantissaBitCnt<float64_t>());
+        const uint64_t mantissa = (uint64_t(ieee754::getMantissaMask<AsFloat>()) & asUint) << (ieee754::getMantissaBitCnt<float64_t>() - ieee754::getMantissaBitCnt<AsFloat>());

```

Comment:
shouldn't you have an `ieee754::extractMantissa(val)` for this?

Reply:
yes, already corrected localy